File: js/src/v2/crypto/aead.ts
/**
 * AUN E2EE V2: AES-256-GCM AEAD
 *
 * 规范引用: §3.3 / §5.1
 * - key:   32 字节
 * - nonce: 12 字节（不可重用）
 * - tag:   16 字节
 *
 * 浏览器实现：使用 WebCrypto subtle.encrypt/decrypt('AES-GCM').
 * - WebCrypto 输出 ciphertext || tag(16B)，与 Python cryptography / Node crypto 一致。
 * - 解密时需要将 ciphertext 与 tag 拼接成单一 buffer 传入 decrypt。
 */

const TAG_LENGTH = 16;

/**
 * AES-256-GCM 加密。
 *
 * @returns { ciphertext, tag } — 分离形式
 */
export async function aesGcmEncrypt(
  key: Uint8Array,
  nonce: Uint8Array,
  plaintext: Uint8Array,
  aad: Uint8Array,
): Promise<{ ciphertext: Uint8Array; tag: Uint8Array }> {
  if (key.length !== 32) throw new Error(`AES-256-GCM key must be 32 bytes, got ${key.length}`);
  if (nonce.length !== 12) throw new Error(`AES-GCM nonce must be 12 bytes, got ${nonce.length}`);

  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    key.slice().buffer,
    { name: 'AES-GCM' },
    false,
    ['encrypt'],
  );
  const ctWithTag = await crypto.subtle.encrypt(
    {
      name: 'AES-GCM',
      iv: nonce.slice().buffer,
      additionalData: aad.slice().buffer,
      tagLength: TAG_LENGTH * 8,
    },
    cryptoKey,
    plaintext.slice().buffer,
  );

  const buf = new Uint8Array(ctWithTag);
  if (buf.length < TAG_LENGTH) {
    throw new Error(`AES-GCM output too short: ${buf.length}`);
  }
  // copy 出独立缓冲区，避免对外暴露 subarray 视图导致后续 slice 行为出乎意料
  const ciphertext = buf.slice(0, buf.length - TAG_LENGTH);
  const tag = buf.slice(buf.length - TAG_LENGTH);
  return { ciphertext, tag };
}

/**
 * AES-256-GCM 解密。
 */
export async function aesGcmDecrypt(
  key: Uint8Array,
  nonce: Uint8Array,
  ciphertext: Uint8Array,
  tag: Uint8Array,
  aad: Uint8Array,
): Promise<Uint8Array> {
  if (key.length !== 32) throw new Error(`AES-256-GCM key must be 32 bytes, got ${key.length}`);
  if (nonce.length !== 12) throw new Error(`AES-GCM nonce must be 12 bytes, got ${nonce.length}`);
  if (tag.length !== TAG_LENGTH) {
    throw new Error(`AES-GCM tag must be 16 bytes, got ${tag.length}`);
  }

  const ctWithTag = new Uint8Array(ciphertext.length + tag.length);
  ctWithTag.set(ciphertext, 0);
  ctWithTag.set(tag, ciphertext.length);

  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    key.slice().buffer,
    { name: 'AES-GCM' },
    false,
    ['decrypt'],
  );
  const pt = await crypto.subtle.decrypt(
    {
      name: 'AES-GCM',
      iv: nonce.slice().buffer,
      additionalData: aad.slice().buffer,
      tagLength: TAG_LENGTH * 8,
    },
    cryptoKey,
    ctWithTag.buffer,
  );
  return new Uint8Array(pt);
}

File: go/v2/crypto/aead.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"fmt"
)

const (
	// AESGCMKeyLen AES-256-GCM 密钥长度（32 字节）
	AESGCMKeyLen = 32
	// AESGCMNonceLen GCM nonce 长度（12 字节）
	AESGCMNonceLen = 12
	// AESGCMTagLen GCM tag 长度（16 字节）
	AESGCMTagLen = 16
)

// AESGCMEncrypt 使用 AES-256-GCM 加密 plaintext。
//
// 返回 (ciphertext, tag)；tag 始终为 16 字节。Go 的 cipher.AEAD.Seal 输出
// `ciphertext || tag`，本函数将其拆分以方便上层组装 wrapped_key 等结构。
func AESGCMEncrypt(key, nonce, plaintext, aad []byte) (ciphertext, tag []byte, err error) {
	if len(key) != AESGCMKeyLen {
		return nil, nil, fmt.Errorf("AEAD: 密钥长度无效，期望 %d 字节，实际 %d", AESGCMKeyLen, len(key))
	}
	if len(nonce) != AESGCMNonceLen {
		return nil, nil, fmt.Errorf("AEAD: nonce 长度无效，期望 %d 字节，实际 %d", AESGCMNonceLen, len(nonce))
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, nil, fmt.Errorf("AEAD: 创建 AES cipher 失败: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, nil, fmt.Errorf("AEAD: 创建 GCM 失败: %w", err)
	}
	sealed := gcm.Seal(nil, nonce, plaintext, aad)
	if len(sealed) < AESGCMTagLen {
		return nil, nil, fmt.Errorf("AEAD: 输出长度异常 %d < tag(%d)", len(sealed), AESGCMTagLen)
	}
	n := len(sealed) - AESGCMTagLen
	// 复制为独立切片，避免上层共享底层数组引发不可见修改
	ct := make([]byte, n)
	copy(ct, sealed[:n])
	tg := make([]byte, AESGCMTagLen)
	copy(tg, sealed[n:])
	return ct, tg, nil
}

// AESGCMDecrypt 使用 AES-256-GCM 解密并校验 tag。
//
// ciphertext 与 tag 必须分别传入；tag 校验失败返回错误。
func AESGCMDecrypt(key, nonce, ciphertext, tag, aad []byte) ([]byte, error) {
	if len(key) != AESGCMKeyLen {
		return nil, fmt.Errorf("AEAD: 密钥长度无效，期望 %d 字节，实际 %d", AESGCMKeyLen, len(key))
	}
	if len(nonce) != AESGCMNonceLen {
		return nil, fmt.Errorf("AEAD: nonce 长度无效，期望 %d 字节，实际 %d", AESGCMNonceLen, len(nonce))
	}
	if len(tag) != AESGCMTagLen {
		return nil, fmt.Errorf("AEAD: tag 长度无效，期望 %d 字节，实际 %d", AESGCMTagLen, len(tag))
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("AEAD: 创建 AES cipher 失败: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("AEAD: 创建 GCM 失败: %w", err)
	}
	sealed := make([]byte, 0, len(ciphertext)+len(tag))
	sealed = append(sealed, ciphertext...)
	sealed = append(sealed, tag...)
	plaintext, err := gcm.Open(nil, nonce, sealed, aad)
	if err != nil {
		return nil, fmt.Errorf("AEAD: 解密/校验失败: %w", err)
	}
	return plaintext, nil
}
